safir/src/safir/uws/_service.py

```diff
@@ -0,0 +1,601 @@
+            callback), or `None` if the execution duration of the job was
+            not changed.
+
+        Raises
+        ------
+        safir.uws._exceptions.PermissionDeniedError
+            If the job ID doesn't exist or is for a user other than the
+            provided user.
+        """
+        job = await self._storage.get(job_id)
+        if job.owner != user:
+            raise PermissionDeniedError(f"Access to job {job_id} denied")
+
+        # Validate the new value.
+        if validator := self._config.validate_execution_duration:
+            duration = validator(duration, job)
+        if duration > self._config.execution_duration:
+            duration = self._config.execution_duration
+
+        # Update the duration in the job.
+        if duration == job.execution_duration:
+            return None
+        await self._storage.update_execution_duration(job_id, duration)
+        if duration.total_seconds() > 0:
+            duration_str = f"{duration.total_seconds()}s"
+        else:
+            duration_str = "unlimited"
+        self._logger.info(
+            "Changed job execution duration",
+            user=user,
+            job_id=job_id,
+            duration=duration_str,
+        )
+        return duration
+
+    def _build_logger_for_job(
+        self, job: UWSJob, params: ParametersModel | None
+    ) -> BoundLogger:
+        """Construct a logger with bound information for a job.
+
+        Parameters
+        ----------
+        job
+            Job for which to report messages.
+        params
+            Job parameters in model form, if available.
+
+        Returns
+        -------
+        BoundLogger
+            Logger with more bound metadata.
+        """
+        logger = self._logger.bind(user=job.owner, job_id=job.job_id)
+        if job.run_id:
+            logger = logger.bind(run_id=job.run_id)
+        if params:
+            logger = logger.bind(parameters=params.model_dump(mode="json"))
+        return logger
+
+    def _validate_parameters(
+        self, params: list[UWSJobParameter]
+    ) -> ParametersModel:
+        """Convert UWS job parameters to the parameter model for the service.
+
+        As a side effect, this also verifies that the parameters are valid,
+        so it is used when creating a job or modifying its parameters to
+        ensure that the new parameters are valid.
+
+        Parameters
+        ----------
+        params
+            Job parameters in the UWS job parameter format.
+
+        Returns
+        -------
+        pydantic.BaseModel
+            Paramters in the model provided by the service, which will be
+            some subclass of `pydantic.BaseModel`.
+
+        Raises
+        ------
+        safir.uws.UWSError
+            Raised if there is some problem with the job parameters.
+        """
+        return self._config.parameters_type.from_job_parameters(params)
+
+    async def _wait_for_job(
```

Comment:
While this is perfectly fine, would it make sense to add a max delay to make sure it doesn't grow too large? Also perhaps using async.timeout here?
I haven't tested this, but here's the idea:

```
async def _wait_for_job(
    self, job: UWSJob, until_not: set[ExecutionPhase], timeout: timedelta
) -> UWSJob:
    async def check_job():
        nonlocal job
        delay = 0.1
        max_delay = 5.0 
        increase_factor = 1.5
        
        while job.phase in until_not:
            await asyncio.sleep(delay)
            job = await self._storage.get(job.job_id)
            delay = min(delay * increase_factor, max_delay)

    with suppress(asyncio.TimeoutError):
        await asyncio.wait_for(check_job(), timeout.total_seconds())

    return job

```

Reply:
This is a great point, thank you! I don't know why I didn't think to use `asyncio.timeout` for this. I've adopted a variation of this approach; it's a lot more compact and straightforward.